### PR TITLE
Merge method, type and const object safety checks

### PR DESCRIPTION
**Pull request by oli-obk**
cc @spastorino and @compiler-errors on the first commit. I believe it to be correct, as the field is only `Some` for assoc types, so just checking the field without checking the assoc kind to be `Type` is fine.

The second commit avoids going through all associated items thrice and just goes over all of them once, running the object safety checks per assoc item kind.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 1]
```diff
@@ -1643,6 +1643,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
+        // `dyn Trait<Assoc = Foo>` desugars to (not RUst syntax) `dyn Trait where <Self as Trait>::Assoc = Foo`.
```
Comment (Nilstrieb) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1646]: typo

**Comment by spastorino** [comment]
@oli-obk the first commit is perfectly right. Those are left overs from when we've added the method to item.

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 2]
```diff
@@ -401,34 +373,52 @@ fn generics_require_sized_self(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
+                _ => item.ident(tcx).span,
+            };
+
+            ObjectSafetyViolation::Method(item.name, v, span)
+        }),
+        // Associated types can only be object safe if they have `Self: Sized` bounds.
+        ty::AssocKind::Type => {
+            if !tcx.features().generic_associated_types_extended {
+                if !tcx.generics_of(item.def_id).params.is_empty() {
+                    if item.opt_rpitit_info.is_none() {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:412]: rightward drift, can you && these?

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 3]
```diff
@@ -401,34 +373,52 @@ fn generics_require_sized_self(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
-        let node = tcx.hir().get_if_local(method.def_id);
-        let span = match (&v, node) {
-            (MethodViolationCode::ReferencesSelfInput(Some(span)), _) => *span,
-            (MethodViolationCode::UndispatchableReceiver(Some(span)), _) => *span,
-            (MethodViolationCode::ReferencesImplTraitInTrait(span), _) => *span,
-            (MethodViolationCode::ReferencesSelfOutput, Some(node)) => {
-                node.fn_decl().map_or(method.ident(tcx).span, |decl| decl.output.span())
+    match item.kind {
+        ty::AssocKind::Const => {
+            Some(ObjectSafetyViolation::AssocConst(item.name, item.ident(tcx).span))
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:391]: Maybe comment saying assoc consts are never object safe

**Review by compiler-errors — APPROVED** [review]
r=me nits or not

**Pull request merged**